jobs/client-regeneration/client_regeneration/main.py

```diff
@@ -0,0 +1,383 @@
+        FROM mozdata.fenix.baseline_clients_last_seen l
+        LEFT JOIN `mozdata.analysis.regen_sim_client_replacements_{str(seed)}` r
+        ON (l.client_id = r.client_id) AND (l.submission_date BETWEEN r.regen_date AND r.regened_last_date)
+        -- now join onto the combined histories for the 27 days following the replacement date
+        LEFT JOIN combined_history c
+        ON (l.client_id = c.original_client_id) AND (l.submission_date = c.submission_date)
+        WHERE l.submission_date >= DATE("{start_date}")
+    """
+    job = client.query(q)
+    job.result()
+    return table_name
+
+
+def create_replacements(
+    client: bigquery.Client,
+    seed: int,
+    start_date: str,
+    end_date: str,
+    column_list: list,
+    lookback: int = 7,
+    use_existing: bool = False,
+):
+    # create a table mapping regenerated clients to their matching replacements.
+    # TODO: Can we get rid of these?
+    replacement_table_name = (
+        """mozdata.analysis.regen_sim_client_replacements_{}""".format(str(seed))
+    )
+    churn_table_name = """mozdata.analysis.regen_sim_churn_pool_{}""".format(str(seed))
+
+    if not use_existing:
+        init_replacement_table(client, seed)
+        init_regen_pool(client, seed, start_date)
+        init_churn_pool(client, seed, start_date, lookback)
+
+    start_dt = datetime.strptime(start_date, "%Y-%m-%d").date()
+    end_dt = datetime.strptime(end_date, "%Y-%m-%d").date()
+    one_day = timedelta(days=1)
+
+    current_dt = start_dt
+    while current_dt <= end_dt:
+        # get the replacements
+        print("""replacing on date {}""".format(str(current_dt)))
+        replacements = (
+            sample_for_replacement_bq(  # TODO: Is this supposed to return something?
+                str(current_dt), column_list, seed, lookback
+            )
+        )
+        print("updating churn pool")
+        update_churn_pool(client, seed, current_dt)
+        current_dt += one_day
+
+
+def run_simulation(
+    client: bigquery.Client,
+    seed: int,
+    start_date: str,
+    column_list: list,
+    end_date: str,
+    lookback: int = 7,
+):
+    # at a high level there are two main steps here 1. go day by day and match regenerated client_ids to replacement
+    # client_ids that "look like" they churned in the prior `lookback` days. write the matches to a table 2. using
+    # the matches from 2, write alternative client histories where regenerated clients are given their replacement ids.
+
+    # this part seems to be working now.
+    create_replacements(
+        client,
+        seed=seed,
+        start_date=start_date,
+        end_date=end_date,
+        column_list=column_list,
+        lookback=lookback,
+    )
+    # next part isn't done yet  # TODO: This works now right?
+    write_attributed_clients_history(client, seed=seed, start_date=start_date)
+    write_usage_history(client, seed=seed, start_date=start_date)
```

Comment:
these are likely to change, we can comment them for now.

```suggestion
    # write_attributed_clients_history(client, seed=seed, start_date=start_date)
    # write_usage_history(client, seed=seed, start_date=start_date)
```